neo/SmartContract/InteropService.NEO.cs

```diff
@@ -226,45 +227,57 @@ private static bool Contract_Create(ApplicationEngine engine)
             byte[] script = engine.CurrentContext.EvaluationStack.Pop().GetByteArray();
             if (script.Length > 1024 * 1024) return false;
-            ContractPropertyState contract_properties = (ContractPropertyState)(byte)engine.CurrentContext.EvaluationStack.Pop().GetBigInteger();
+
+            var manifest = engine.CurrentContext.EvaluationStack.Pop().GetString();
+            if (manifest.Length > ContractManifest.MaxLength) return false;
+
             UInt160 hash = script.ToScriptHash();
             ContractState contract = engine.Snapshot.Contracts.TryGet(hash);
-            if (contract == null)
+            if (contract != null) return false;
+            contract = new ContractState
             {
-                contract = new ContractState
-                {
-                    Script = script,
-                    ContractProperties = contract_properties
-                };
-                engine.Snapshot.Contracts.Add(hash, contract);
-            }
+                Script = script,
+                Manifest = ContractManifest.Parse(manifest)
+            };
+
+            if (!contract.Manifest.IsValid(hash)) return false;
+
+            engine.Snapshot.Contracts.Add(hash, contract);
             engine.CurrentContext.EvaluationStack.Push(StackItem.FromInterface(contract));
             return true;
         }
 
-        private static bool Contract_Migrate(ApplicationEngine engine)
+        private static bool Contract_Update(ApplicationEngine engine)
         {
             if (engine.Trigger != TriggerType.Application) return false;
+
             byte[] script = engine.CurrentContext.EvaluationStack.Pop().GetByteArray();
             if (script.Length > 1024 * 1024) return false;
-            ContractPropertyState contract_properties = (ContractPropertyState)(byte)engine.CurrentContext.EvaluationStack.Pop().GetBigInteger();
-            UInt160 hash = script.ToScriptHash();
-            ContractState contract = engine.Snapshot.Contracts.TryGet(hash);
-            if (contract == null)
+            var manifest = engine.CurrentContext.EvaluationStack.Pop().GetString();
+            if (manifest.Length > ContractManifest.MaxLength) return false;
+
+            var contract = engine.Snapshot.Contracts.TryGet(engine.CurrentScriptHash);
+            if (contract is null) return false;
+
+            if (script.Length > 0)
             {
+                UInt160 hash_new = script.ToScriptHash();
+                if (hash_new.Equals(engine.CurrentScriptHash)) return false;
+                if (engine.Snapshot.Contracts.TryGet(hash_new) != null) return false;
                 contract = new ContractState
                 {
                     Script = script,
-                    ContractProperties = contract_properties
+                    Manifest = contract.Manifest
                 };
-                engine.Snapshot.Contracts.Add(hash, contract);
+                contract.Manifest.Abi.Hash = hash_new;
```

Comment:
If someone provide both, script and manifest, and the manifest with the wrong hash, we calculate it, but .... is this right?

Reply:
We calculate for the old one.

Reply:
I think it is right, if you have the script you should calculate the Hash, but if it is different maybe we could throw an exception and fail the transaction.